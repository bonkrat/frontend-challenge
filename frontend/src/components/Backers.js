import React from "react";
import ReactTooltip from "react-tooltip";
import * as styles from "./Backers.styles";
import PropTypes from "prop-types";

/**
 * Renders a list of backers.
 */
const Backers = ({ backers, size }) => (
  <div className={styles.backers}>
    {backers.map(({ id, name, avatar }) => (
      <React.Fragment key={id}>
        <ReactTooltip effect="solid" />
        <div
          data-tip={name}
          className={styles.backer(size)}
          style={{
            backgroundImage: `url("http://localhost:4000/avatars/${avatar}")`,
            backgroundPosition: "center",
            backgroundSize: "cover"
          }}
        ></div>
      </React.Fragment>
    ))}
  </div>
);

Backers.defaultProps = {
  backers: [],
  size: 6
};

Backers.propTypes = {
  /**
   * An array of backers to render.
   */
  backers: PropTypes.array,
  /**
   * The size of the avatars (1, 2, 4, 6, 8, 16...)
   */
  size: PropTypes.number
};

export default Backers;
