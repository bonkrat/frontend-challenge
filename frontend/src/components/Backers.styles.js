import { classnames } from "tailwindcss-classnames";

export const backers = classnames("flex", "text-base", "text-gray-500");

export const backer = size =>
  classnames(
    "rounded-full",
    `h-${size}`,
    `w-${size}`,
    "flex",
    "items-center",
    "justify-center",
    "text-white"
  );
