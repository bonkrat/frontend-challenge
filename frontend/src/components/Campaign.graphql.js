import { gql } from "apollo-boost";

export const CAMPAIGN_QUERY = gql`
  query getCampaign($id: ID!) {
    campaign(id: $id) {
      id
      title
      creator {
        name
      }
      description
      backers {
        id
        name
        avatar
      }
      image
      funds
      goal
    }
  }
`;
