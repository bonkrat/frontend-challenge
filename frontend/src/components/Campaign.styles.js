import { classnames } from "tailwindcss-classnames";

export const campaignHeader = classnames(
  "flex",
  "flex-col",
  "bg-purple-100",
  "w-full",
  //   "lg:px-16",
  "p-8",
  "flex",
  "justify-center",
  "items-center"
);

export const campaignTitle = classnames(
  "text-2xl",
  "text-white",
  "text-indigo-800"
);

export const campaignCreator = classnames("text-md", "text-purple-400");

export const campaignBody = classnames(
  "flex",
  "flex-auto",
  "flex-wrap",
  "justify-center",
  "h-auto"
);

export const image = classnames(
  "h-auto",
  "lg:flex-1",
  "md:flex-none",
  "sm: flex-none",
  "bg-cover",
  "rounded-br-lg",
  "text-center",
  "overflow-hidden"
);

export const campaignDetails = classnames(
  "h-auto",
  "flex",
  "flex-col",
  "justify-between",
  "md:flex-1",
  "py-4",
  "px-8"
);

export const goal = classnames(
  "text-lg",
  "text-gray-00",
  "flex",
  "justify-between"
);

export const funds = classnames("text-purple-900");

export const fundBar = classnames("w-full", "text-2xl");

export const campaignDescription = classnames("my-10", "lg:mb-16");

export const backedBy = classnames("md: mb-16", "my-4");

export const backedByText = classnames("mb-2", "text-gray-700");
