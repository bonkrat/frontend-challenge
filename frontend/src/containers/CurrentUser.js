import { useState } from "react";
import { createContainer } from "unstated-next";

export const CurrentUser = () => {
  const [user, setUser] = useState({});
  return { user, setUser };
};

/**
 * Creates a global state container with the current logged in user.
 */
export default createContainer(CurrentUser);
