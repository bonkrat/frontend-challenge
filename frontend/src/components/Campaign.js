import React from "react";
import PropTypes from "prop-types";
import * as styles from "./Campaign.styles";
import useQuery from "../hooks/useQuery";
import { CAMPAIGN_QUERY } from "./Campaign.graphql";
import FundedBar from "./FundedBar";
import FundForm from "./FundForm";
import Backers from "./Backers";
import CampaignPlaceholder from "./CampaignPlaceholder";

/**
 * Renders a fullpage view of a campaign.
 */
export const Campaign = ({
  id,
  title,
  creator,
  image,
  goal,
  funds,
  description,
  backers
}) => (
  <React.Fragment>
    <div className={styles.campaignHeader}>
      <div className={styles.campaignTitle}>{title}</div>
      <div className={styles.campaignCreator}>Created by {creator.name}</div>
    </div>
    <div className={styles.campaignBody}>
      <img
        alt={title}
        className={styles.image}
        src={`http://localhost:4000/images/${image}`}
      />
      <div className={styles.campaignDetails}>
        <FundedBar className={styles.fundBar} funds={funds} goal={goal} />
        <div className={styles.goal}>
          <span className={styles.funds}>{funds < goal && `$${funds}`}</span>
          <span>${funds < goal ? goal : funds}</span>
        </div>
        <p className={styles.campaignDescription}>{description}</p>
        {backers.length > 0 && (
          <div className={styles.backedBy}>
            <div className={styles.backedByText}>Backed by:</div>
            <Backers backers={backers} size={12} />
          </div>
        )}
        <FundForm campaignId={id} />
      </div>
    </div>
  </React.Fragment>
);

Campaign.propTypes = {
  /**
   * The id of the campaign.
   */
  id: PropTypes.string,
  /**
   * The title of the campaign.
   */
  title: PropTypes.string,
  /**
   * The creator of the campaign
   */
  creator: PropTypes.object,
  /**
   * The URL of the image of the campaign.
   */
  image: PropTypes.string,
  /**
   * The goal to reach for the campaign.
   */
  goal: PropTypes.number,
  /**
   * The funds the campaign has raised.
   */
  funds: PropTypes.number,
  /**
   * The description of the campaign.
   */
  description: PropTypes.string,
  /**
   * The users that have backed the campaign.
   */
  backers: PropTypes.array
};

/**
 * Queries for and renders a campaign.
 */
export const CampaignQuery = ({ id }) => {
  const { loading, error, data: { campaign } = {} } = useQuery(
    "CAMPAIGN_VIEW",
    CAMPAIGN_QUERY,
    {
      variables: {
        id
      }
    }
  );

  if (error || loading) return <CampaignPlaceholder />;

  return <Campaign {...campaign} />;
};

CampaignQuery.propTypes = {
  /**
   * The id of the campaign to query and render.
   */
  id: PropTypes.string
};

export default CampaignQuery;
