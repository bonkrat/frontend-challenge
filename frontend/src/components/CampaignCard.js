import React from "react";
import PropTypes from "prop-types";
import FundedBar from "./FundedBar";
import FundButton from "./FundButton";
import Backers from "./Backers";
import { Link } from "react-router-dom";
import styles from "./CampaignCard.styles";

/**
 * Links children to a product
 */
export const LinkToProduct = ({ id, style, children }) => (
  <Link to={"/" + id} style={style}>
    {children}
  </Link>
);

LinkToProduct.propTypes = {
  /**
   * The ID of the product to link to.
   */
  id: PropTypes.string,
  /**
   * The style of the link.
   */
  style: PropTypes.object
};

/**
 * A card rendering a campaign.
 *
 * @param {object} props
 */
const CampaignCard = ({
  size = ["flex-1"],
  hero,
  id,
  title,
  description,
  image,
  goal,
  funds,
  creator,
  backers
}) => (
  <div className={styles.card(size)}>
    <LinkToProduct
      id={id}
      style={{
        backgroundImage: `url("http://localhost:4000/images/${image}")`,
        backgroundPosition: "center",
        backgroundSize: "cover"
      }}
    >
      <div className={styles.image(hero)} title="name"></div>
    </LinkToProduct>
    <div className={styles.cardContents(hero)}>
      <div className={styles.title}>
        <LinkToProduct id={id}>
          <div className={styles.titleText}>{title}</div>
        </LinkToProduct>
        {creator && (
          <div className={styles.creator(hero)}>Created by {creator.name}</div>
        )}
        <p className={styles.description(hero)}>{description}</p>
      </div>
      <Backers backers={backers} />
      <div className={styles.fund}>
        <FundedBar className="w-3/4" goal={goal} funds={funds} />
        <LinkToProduct id={id}>
          <FundButton className="w-1/4" />
        </LinkToProduct>
      </div>
    </div>
  </div>
);

CampaignCard.propTypes = {
  /**
   * An array of tailwind classes describing the size of the card.
   */
  size: PropTypes.array,
  /**
   * If true, the card is rendered larger as a 'hero' card.
   */
  hero: PropTypes.bool,
  /**
   * The id of the campaign.
   */
  id: PropTypes.string,
  /**
   * The title of the campaign.
   */
  title: PropTypes.string,
  /**
   * The description of the campaign.
   */
  description: PropTypes.string,
  /**
   * The URL to the image for the campaign.
   */
  image: PropTypes.string,
  /**
   * The goal amount of the campaign.
   */
  goal: PropTypes.number,
  /**
   * The amount of funding the campaign has raised.
   */
  funds: PropTypes.number,
  /**
   * The creator of the campaign.
   */
  creator: PropTypes.object,
  /**
   * Backers of the campaign.
   */
  backers: PropTypes.array
};

export default CampaignCard;
