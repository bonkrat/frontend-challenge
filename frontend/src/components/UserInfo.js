import React from "react";
import PropTypes from "prop-types";
import * as styles from "./UserInfo.styles";
import { USER_INFO_QUERY } from "./UserInfo.graphql";
import useQuery from "../hooks/useQuery";
import CurrentUser from "../containers/CurrentUser";
import ReactTooltip from "react-tooltip";

/**
 * Renders the logged in user's info
 */
export const UserInfo = ({ loading, name = "", avatar, balance }) => (
  <div className={styles.userInfo}>
    {balance && <div className={styles.balance}>${balance}</div>}
    <ReactTooltip effect="solid" place="bottom" />
    <div
      data-tip={name}
      className={styles.avatar(loading)}
      style={{
        backgroundImage: `url("http://localhost:4000/avatars/${avatar}")`,
        backgroundPosition: "center",
        backgroundSize: "cover"
      }}
    ></div>
  </div>
);

UserInfo.propTypes = {
  /**
   * Renders plcaeholder if loading
   */
  loading: PropTypes.bool,
  /**
   * The user's name
   */
  name: PropTypes.string,
  /**
   * URL to the users avatar
   */
  avatar: PropTypes.string,
  /**
   * The user's balance
   */
  balance: PropTypes.number
};

/**
 * Queries and renders the User's info.
 */
export const UserInfoQuery = () => {
  const { loading, error, data = {} } = useQuery("USER_INFO", USER_INFO_QUERY);
  const { setUser } = CurrentUser.useContainer();

  if (error) return <div></div>;
  setUser(data.loggedInUser);

  return <UserInfo loading={loading} {...data.loggedInUser} />;
};

export default UserInfoQuery;
