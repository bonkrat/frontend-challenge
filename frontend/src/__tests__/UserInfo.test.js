import React from "react";
import { shallow } from "enzyme";
import { UserInfoQuery, UserInfo } from "../components/UserInfo";
import toJson from "enzyme-to-json";
import useQuery from "../hooks/useQuery";
import { USER_INFO_QUERY } from "../components/UserInfo.graphql";
import CurrentUser from "../containers/CurrentUser";

jest.mock("../containers/CurrentUser", () => ({
  useContainer: jest.fn(() => ({
    setUser: jest.fn()
  }))
}));

jest.mock("../hooks/useQuery", () =>
  jest
    .fn(() => ({
      loading: false,
      error: false,
      data: {}
    }))
    .mockReturnValueOnce({
      loading: true,
      error: false,
      data: {}
    })
    .mockReturnValueOnce({
      loading: false,
      error: true,
      data: {}
    })
    .mockReturnValueOnce({
      loading: false,
      error: false,
      data: {
        loggedInUser: {
          id: "foo"
        }
      }
    })
);

let wrapper;

describe("UserInfo", () => {
  const props = {
    loading: true,
    name: "foo",
    avatar: "foo.jpg",
    balance: 9001
  };

  beforeEach(() => {
    wrapper = shallow(<UserInfo {...props} />);
  });

  it("renders correctly", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("renders loading styles", () => {
    wrapper.setProps({ loading: true });
    wrapper.update();

    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

describe("UserInfoQuery", () => {
  beforeEach(() => {
    wrapper = shallow(<UserInfoQuery />);
  });

  afterEach(() => {
    CurrentUser.useContainer.mockClear();
  });

  it("sets loading on user info", () => {
    expect(wrapper.find(UserInfo).prop("loading")).toBe(true);
  });

  it("returns nothing if there's an error", () => {
    expect(wrapper.find(UserInfo).exists()).toBeFalsy();
  });

  it("queries for user info and sets it to user info", () => {
    expect(useQuery).toHaveBeenCalledWith("USER_INFO", USER_INFO_QUERY);
    expect(wrapper.find(UserInfo).prop("id")).toBe("foo");
  });

  it("sets the user to global state", () => {
    expect(CurrentUser.useContainer).toHaveBeenCalled();
  });
});
