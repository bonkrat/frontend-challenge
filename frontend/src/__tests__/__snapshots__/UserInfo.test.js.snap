// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserInfo renders correctly 1`] = `
<div
  className="flex items-center"
>
  <div
    className="text-purple-800 text-white font-bold rounded px-4 py-2"
  >
    $
    9001
  </div>
  <ReactTooltip
    clickable={false}
    effect="solid"
    insecure={true}
    place="bottom"
    resizeHide={true}
    wrapper="div"
  />
  <div
    className="rounded-full h-12 w-12 flex items-center justify-center text-white border"
    data-tip="foo"
    style={
      Object {
        "backgroundImage": "url(\\"http://localhost:4000/avatars/foo.jpg\\")",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  />
</div>
`;

exports[`UserInfo renders loading styles 1`] = `
<div
  className="flex items-center"
>
  <div
    className="text-purple-800 text-white font-bold rounded px-4 py-2"
  >
    $
    9001
  </div>
  <ReactTooltip
    clickable={false}
    effect="solid"
    insecure={true}
    place="bottom"
    resizeHide={true}
    wrapper="div"
  />
  <div
    className="rounded-full h-12 w-12 flex items-center justify-center text-white border"
    data-tip="foo"
    style={
      Object {
        "backgroundImage": "url(\\"http://localhost:4000/avatars/foo.jpg\\")",
        "backgroundPosition": "center",
        "backgroundSize": "cover",
      }
    }
  />
</div>
`;
